controller/lib/comms/comms_nanopb.cpp

```diff
@@ -0,0 +1,90 @@
+#ifdef TEST_MODE
+#include "command_handlers-mock.h"
+#include "serialIO-mock.h"
+#else
+#include "command_handlers.h"
+#include "serialIO.h"
+#include <Arduino.h>
+#endif
+
+#include "comms.h"
+#include "network_protocol.pb.h"
+#include "serdes.h"
+// #include "version.h"
+#include "network_protocol.pb.h"
+
+#define PACKET_LEN_MAX (32)
+static uint8_t tx_buffer[PACKET_LEN_MAX];
+static uint16_t tx_data_length;
+static bool output_buffer_ready = false;
+
+static uint8_t rx_buffer[PACKET_LEN_MAX];
+// static uint16_t rx_data_length;
+static uint8_t rx_idx = 0;
+static uint64_t last_rx = 0;
+static uint8_t rx_in_progress = false;
+
+#define RX_TIMEOUT 1
+
+void comms_init() { serialIO_init(); }
+
+inline static bool is_time_process_packet() {
+  return millis() - last_rx > RX_TIMEOUT;
+}
+
+void comms_sendResetState() {
+  // TODO
+}
+
+void comms_sendPeriodicReadings(float pressure, float volume, float flow) {
+  // TODO solve tx overflow
+  if (output_buffer_ready) {
+    return;
+  }
+  uint64_t time;
+  time = millis();
+  bool status =
+      serdes_encode_status_packet(time, pressure, volume, flow, tx_buffer,
+                                  PACKET_LEN_MAX, (size_t *)&tx_data_length);
```

Comment:
What is UB? :)

Since we are on ARM now, all variables should move to int32. Should we use size_t for all size/length variables?

Reply:
UB == "undefined behavior".  Meaning, from the compiler's perspective, you have violated one of the rules of correct C++ programming, therefore the compiler is released from any and all obligation to create a correct program for you.

> Since we are on ARM now, all variables should move to int32. Should we use size_t for all size/length variables?

size_t is the data type used by the C++ standard library for lengths.  I think matching the STL is pretty much the only reason to use it.  Since we're not using STL collections (all of them require `malloc`, which we're not using), I think using explicit-width types (e.g. `int32`) is probably the way to go.  There's no downside and it's a lot safer/more explicit.

Reply:
I also use explicit types always, but this time `size_t` comes from nanopb. I feel it will be uint32 on ARM.